File: clients/go/consensus/spend_verify.go
package consensus

func extractCryptoSigAndSighash(w WitnessItem) ([]byte, uint8, error) {
	if len(w.Signature) == 0 {
		return nil, 0, txerr(TX_ERR_PARSE, "missing sighash_type byte")
	}
	sighashType := w.Signature[len(w.Signature)-1]
	if !IsValidSighashType(sighashType) {
		return nil, 0, txerr(TX_ERR_SIGHASH_TYPE_INVALID, "invalid sighash_type")
	}
	return w.Signature[:len(w.Signature)-1], sighashType, nil
}

// extractSigAndDigest extracts the cryptographic signature bytes and computes
// the sighash digest from a witness item.
func extractSigAndDigest(w WitnessItem, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte) ([]byte, [32]byte, error) {
	return extractSigAndDigestWithCache(w, tx, inputIndex, inputValue, chainID, nil)
}

func extractSigAndDigestWithCache(w WitnessItem, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, cache *SighashV1PrehashCache) ([]byte, [32]byte, error) {
	cryptoSig, sighashType, err := extractCryptoSigAndSighash(w)
	if err != nil {
		return nil, [32]byte{}, err
	}
	var digest [32]byte
	if cache != nil {
		digest, err = SighashV1DigestWithCache(cache, inputIndex, inputValue, chainID, sighashType)
	} else {
		digest, err = SighashV1DigestWithType(tx, inputIndex, inputValue, chainID, sighashType)
	}
	if err != nil {
		return nil, [32]byte{}, err
	}
	return cryptoSig, digest, nil
}

// verifyMLDSAKeyAndSig verifies an ML-DSA-87 witness item's key binding and
// cryptographic signature. The caller must validate witness item lengths and
// suite ID before calling this function.
func verifyMLDSAKeyAndSig(w WitnessItem, expectedKeyID [32]byte, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, context string) error {
	return verifyMLDSAKeyAndSigWithCache(w, expectedKeyID, tx, inputIndex, inputValue, chainID, nil, context)
}

func verifyMLDSAKeyAndSigWithCache(w WitnessItem, expectedKeyID [32]byte, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, cache *SighashV1PrehashCache, context string) error {
	return verifyKeyAndSigWithRegistryCache(w, expectedKeyID, spendSigContext{
		tx:         tx,
		inputIndex: inputIndex,
		inputValue: inputValue,
		chainID:    chainID,
		cache:      cache,
		context:    context,
	})
}

type spendSigContext struct {
	tx         *Tx
	inputIndex uint32
	inputValue uint64
	chainID    [32]byte
	cache      *SighashV1PrehashCache
	registry   *SuiteRegistry
	context    string
}

// verifyKeyAndSigWithRegistryCache verifies a witness item's key binding and
// cryptographic signature using registry-aware algorithm dispatch. When
// ctx.registry is nil, the canonical default live registry is used; callers do
// not get a separate implicit legacy verifier path.
func verifyKeyAndSigWithRegistryCache(w WitnessItem, expectedKeyID [32]byte, ctx spendSigContext) error {
	if sha3_256(w.Pubkey) != expectedKeyID {
		return txerr(TX_ERR_SIG_INVALID, ctx.context+" key binding mismatch")
	}
	cryptoSig, digest, err := extractSigAndDigestWithCache(w, ctx.tx, ctx.inputIndex, ctx.inputValue, ctx.chainID, ctx.cache)
	if err != nil {
		return err
	}
	ok, err := verifySigWithRegistry(w.SuiteID, w.Pubkey, cryptoSig, digest, ctx.registry)
	if err != nil {
		return err
	}
	if !ok {
		return txerr(TX_ERR_SIG_INVALID, ctx.context+" signature invalid")
	}
	return nil
}

func validateP2PKSpend(entry UtxoEntry, w WitnessItem, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, blockHeight uint64) error {
	return validateP2PKSpendWithCache(entry, w, tx, inputIndex, inputValue, chainID, blockHeight, nil)
}

func validateP2PKSpendWithCache(entry UtxoEntry, w WitnessItem, tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, blockHeight uint64, cache *SighashV1PrehashCache) error {
	return validateP2PKSpendAtHeight(p2pkSpendCheck{
		entry:       entry,
		witness:     w,
		blockHeight: blockHeight,
		rotation:    DefaultRotationProvider{},
		sig: spendSigContext{
			tx:         tx,
			inputIndex: inputIndex,
			inputValue: inputValue,
			chainID:    chainID,
			cache:      cache,
			registry:   DefaultSuiteRegistry(),
		},
	})
}

type p2pkSpendCheck struct {
	entry       UtxoEntry
	witness     WitnessItem
	sig         spendSigContext
	blockHeight uint64
	rotation    RotationProvider
}

func defaultSpendProviders(rotation RotationProvider, registry *SuiteRegistry) (RotationProvider, *SuiteRegistry) {
	if rotation == nil {
		rotation = DefaultRotationProvider{}
	}
	if registry == nil {
		registry = DefaultSuiteRegistry()
	}
	return rotation, registry
}

// validateP2PKSpendAtHeight validates a P2PK spend using the suite registry
// and rotation provider for suite validation, length checks, and signature
// dispatch. When rotation or registry is nil, defaults are used (ML-DSA-87
// genesis set).
func validateP2PKSpendAtHeight(check p2pkSpendCheck) error {
	rotation, registry := defaultSpendProviders(check.rotation, check.sig.registry)
	w := check.witness
	nativeSpend := rotation.NativeSpendSuites(check.blockHeight)
	if !nativeSpend.Contains(w.SuiteID) {
		return txerr(TX_ERR_SIG_ALG_INVALID, "CORE_P2PK suite not in native spend set")
	}

	params, ok := registry.Lookup(w.SuiteID)
	if !ok {
		return txerr(TX_ERR_SIG_ALG_INVALID, "CORE_P2PK suite not registered")
	}

	if len(w.Pubkey) != params.PubkeyLen || len(w.Signature) != params.SigLen+1 {
		return txerr(TX_ERR_SIG_NONCANONICAL, "non-canonical witness item lengths")
	}

	if len(check.entry.CovenantData) != MAX_P2PK_COVENANT_DATA || check.entry.CovenantData[0] != w.SuiteID {
		return txerr(TX_ERR_COVENANT_TYPE_INVALID, "CORE_P2PK covenant_data invalid")
	}

	var keyID [32]byte
	copy(keyID[:], check.entry.CovenantData[1:33])
	sig := check.sig
	sig.registry = registry
	sig.context = "CORE_P2PK"
	return verifyKeyAndSigWithRegistryCache(w, keyID, sig)
}

type thresholdSigSpendCheck struct {
	keys        [][32]byte
	threshold   uint8
	witnesses   []WitnessItem
	sig         spendSigContext
	blockHeight uint64
	rotation    RotationProvider
}

func validateThresholdSigSpend(check thresholdSigSpendCheck) error {
	if check.rotation == nil {
		check.rotation = DefaultRotationProvider{}
	}
	if check.sig.registry == nil {
		check.sig.registry = DefaultSuiteRegistry()
	}
	return validateThresholdSigSpendAtHeight(check)
}

// validateThresholdSigSpendAtHeight validates a threshold-sig spend using the
// suite registry and rotation provider. When rotation or registry is nil,
// defaults are used (ML-DSA-87 genesis set).
func validateThresholdSigSpendAtHeight(check thresholdSigSpendCheck) error {
	rotation, registry := defaultSpendProviders(check.rotation, check.sig.registry)
	if len(check.witnesses) != len(check.keys) {
		return txerr(TX_ERR_PARSE, "witness slot assignment mismatch")
	}

	nativeSpend := rotation.NativeSpendSuites(check.blockHeight)
	sig := check.sig
	sig.registry = registry
	valid := 0

	for i := range check.keys {
		counted, err := validateThresholdWitness(check.witnesses[i], check.keys[i], nativeSpend, registry, sig)
		if err != nil {
			return err
		}
		if counted {
			valid++
		}
	}

	if valid < int(check.threshold) {
		return txerr(TX_ERR_SIG_INVALID, sig.context+" threshold not met")
	}
	return nil
}

func validateThresholdWitness(w WitnessItem, key [32]byte, nativeSpend *NativeSuiteSet, registry *SuiteRegistry, sig spendSigContext) (bool, error) {
	if sentinel, err := validateSentinelThresholdWitness(w); sentinel || err != nil {
		return false, err
	}

	if !nativeSpend.Contains(w.SuiteID) {
		return false, txerr(TX_ERR_SIG_ALG_INVALID, sig.context+" suite not in native spend set")
	}

	params, ok := registry.Lookup(w.SuiteID)
	if !ok {
		return false, txerr(TX_ERR_SIG_ALG_INVALID, sig.context+" suite not registered")
	}

	if len(w.Pubkey) != params.PubkeyLen || len(w.Signature) != params.SigLen+1 {
		return false, txerr(TX_ERR_SIG_NONCANONICAL, "non-canonical witness item lengths")
	}

	if err := verifyKeyAndSigWithRegistryCache(w, key, sig); err != nil {
		return false, err
	}
	return true, nil
}

func validateSentinelThresholdWitness(w WitnessItem) (bool, error) {
	if w.SuiteID != SUITE_ID_SENTINEL {
		return false, nil
	}
	if len(w.Pubkey) != 0 || len(w.Signature) != 0 {
		return true, txerr(TX_ERR_PARSE, "SENTINEL witness must be keyless")
	}
	return true, nil
}

File: clients/rust/crates/rubin-consensus/src/spend_verify.rs
use crate::constants::{MAX_P2PK_COVENANT_DATA, SUITE_ID_SENTINEL};
use crate::error::{ErrorCode, TxError};
use crate::hash::sha3_256;
use crate::sig_queue::{queue_or_verify_signature, SigCheckQueue};
use crate::sighash::{is_valid_sighash_type, sighash_v1_digest_with_cache, SighashV1PrehashCache};
use crate::suite_registry::{
    DefaultRotationProvider, NativeSuiteSet, RotationProvider, SuiteParams, SuiteRegistry,
};
use crate::tx::Tx;
use crate::tx::WitnessItem;
use crate::utxo_basic::UtxoEntry;

pub(crate) fn extract_crypto_sig_and_sighash(w: &WitnessItem) -> Result<(&[u8], u8), TxError> {
    let Some((&sighash_type, crypto_sig)) = w.signature.split_last() else {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "missing sighash_type byte",
        ));
    };
    if !is_valid_sighash_type(sighash_type) {
        return Err(TxError::new(
            ErrorCode::TxErrSighashTypeInvalid,
            "invalid sighash_type",
        ));
    }
    Ok((crypto_sig, sighash_type))
}

#[allow(dead_code)]
pub(crate) fn validate_p2pk_spend(
    entry: &UtxoEntry,
    w: &WitnessItem,
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
) -> Result<(), TxError> {
    let mut cache = SighashV1PrehashCache::new(tx)?;
    validate_p2pk_spend_with_cache(
        entry,
        w,
        tx,
        input_index,
        input_value,
        chain_id,
        block_height,
        &mut cache,
    )
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_p2pk_spend_with_cache(
    entry: &UtxoEntry,
    w: &WitnessItem,
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    cache: &mut SighashV1PrehashCache<'_>,
) -> Result<(), TxError> {
    validate_p2pk_spend_at_height(
        entry,
        w,
        tx,
        input_index,
        input_value,
        chain_id,
        block_height,
        cache,
        None,
        None,
    )
}

/// Rotation-aware P2PK spend validation. When rotation or registry is None,
/// uses defaults (ML-DSA-87 genesis set). Parity with Go `validateP2PKSpendAtHeight`.
#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_p2pk_spend_at_height(
    entry: &UtxoEntry,
    w: &WitnessItem,
    _tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    cache: &mut SighashV1PrehashCache<'_>,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
) -> Result<(), TxError> {
    validate_p2pk_spend_q(
        entry,
        w,
        input_index,
        input_value,
        chain_id,
        block_height,
        cache,
        None,
        rotation,
        registry,
    )
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_p2pk_spend_q(
    entry: &UtxoEntry,
    w: &WitnessItem,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    cache: &mut SighashV1PrehashCache<'_>,
    sig_queue: Option<&mut SigCheckQueue>,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
) -> Result<(), TxError> {
    let default_rp = DefaultRotationProvider;
    let default_reg = SuiteRegistry::default_registry();
    let rp: &dyn RotationProvider = rotation.unwrap_or(&default_rp);
    let reg = registry.unwrap_or(&default_reg);
    let native_spend = rp.native_spend_suites(block_height);
    let params = spend_suite_params(
        w,
        &native_spend,
        reg,
        "CORE_P2PK suite not in native spend set",
        "CORE_P2PK suite not registered",
    )?;
    validate_witness_item_lengths(w, params)?;
    let key_id = p2pk_covenant_key_id(entry, w.suite_id)?;
    let mut sig_queue = sig_queue;
    verify_mldsa_key_and_sig_q(
        w,
        key_id,
        input_index,
        input_value,
        chain_id,
        cache,
        reg,
        &mut sig_queue,
        TxError::new(ErrorCode::TxErrSigInvalid, "CORE_P2PK key binding mismatch"),
        TxError::new(ErrorCode::TxErrSigInvalid, "CORE_P2PK signature invalid"),
    )
}

fn spend_suite_params<'a>(
    w: &WitnessItem,
    native_spend: &NativeSuiteSet,
    registry: &'a SuiteRegistry,
    suite_not_native_message: &'static str,
    suite_not_registered_message: &'static str,
) -> Result<&'a SuiteParams, TxError> {
    if !native_spend.contains(w.suite_id) {
        return Err(TxError::new(
            ErrorCode::TxErrSigAlgInvalid,
            suite_not_native_message,
        ));
    }
    registry
        .lookup(w.suite_id)
        .ok_or_else(|| TxError::new(ErrorCode::TxErrSigAlgInvalid, suite_not_registered_message))
}

fn validate_witness_item_lengths(w: &WitnessItem, params: &SuiteParams) -> Result<(), TxError> {
    if w.pubkey.len() as u64 == params.pubkey_len && w.signature.len() as u64 == params.sig_len + 1
    {
        return Ok(());
    }
    Err(TxError::new(
        ErrorCode::TxErrSigNoncanonical,
        "non-canonical witness item lengths",
    ))
}

fn p2pk_covenant_key_id(entry: &UtxoEntry, suite_id: u8) -> Result<[u8; 32], TxError> {
    if entry.covenant_data.len() as u64 != MAX_P2PK_COVENANT_DATA
        || entry.covenant_data[0] != suite_id
    {
        return Err(TxError::new(
            ErrorCode::TxErrCovenantTypeInvalid,
            "CORE_P2PK covenant_data invalid",
        ));
    }
    let mut key_id = [0u8; 32];
    key_id.copy_from_slice(&entry.covenant_data[1..33]);
    Ok(key_id)
}

#[allow(dead_code)]
#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_threshold_sig_spend(
    keys: &[[u8; 32]],
    threshold: u8,
    ws: &[WitnessItem],
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    context: &'static str,
) -> Result<(), TxError> {
    let mut cache = SighashV1PrehashCache::new(tx)?;
    validate_threshold_sig_spend_with_cache(
        keys,
        threshold,
        ws,
        tx,
        input_index,
        input_value,
        chain_id,
        block_height,
        context,
        &mut cache,
    )
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_threshold_sig_spend_with_cache(
    keys: &[[u8; 32]],
    threshold: u8,
    ws: &[WitnessItem],
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    context: &'static str,
    cache: &mut SighashV1PrehashCache<'_>,
) -> Result<(), TxError> {
    validate_threshold_sig_spend_at_height(
        keys,
        threshold,
        ws,
        tx,
        input_index,
        input_value,
        chain_id,
        block_height,
        context,
        cache,
        None,
        None,
    )
}

/// Rotation-aware threshold-sig spend validation. When rotation or registry
/// is None, uses defaults (ML-DSA-87 genesis set). Parity with Go
/// `validateThresholdSigSpendAtHeight`.
#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_threshold_sig_spend_at_height(
    keys: &[[u8; 32]],
    threshold: u8,
    ws: &[WitnessItem],
    _tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    context: &'static str,
    cache: &mut SighashV1PrehashCache<'_>,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
) -> Result<(), TxError> {
    validate_threshold_sig_spend_q(
        keys,
        threshold,
        ws,
        input_index,
        input_value,
        chain_id,
        block_height,
        context,
        cache,
        None,
        rotation,
        registry,
    )
}

#[derive(Clone, Copy)]
enum ThresholdSigErrorDetail {
    SuiteNotInNativeSpendSet,
    SuiteNotRegistered,
    KeyBindingMismatch,
    SignatureInvalid,
    ThresholdNotMet,
}

fn threshold_sig_error_message(
    context: &'static str,
    detail: ThresholdSigErrorDetail,
) -> &'static str {
    match (context, detail) {
        ("CORE_MULTISIG", ThresholdSigErrorDetail::SuiteNotInNativeSpendSet) => {
            "CORE_MULTISIG suite not in native spend set"
        }
        ("CORE_MULTISIG", ThresholdSigErrorDetail::SuiteNotRegistered) => {
            "CORE_MULTISIG suite not registered"
        }
        ("CORE_MULTISIG", ThresholdSigErrorDetail::KeyBindingMismatch) => {
            "CORE_MULTISIG key binding mismatch"
        }
        ("CORE_MULTISIG", ThresholdSigErrorDetail::SignatureInvalid) => {
            "CORE_MULTISIG signature invalid"
        }
        ("CORE_MULTISIG", ThresholdSigErrorDetail::ThresholdNotMet) => {
            "CORE_MULTISIG threshold not met"
        }
        ("CORE_VAULT", ThresholdSigErrorDetail::SuiteNotInNativeSpendSet) => {
            "CORE_VAULT suite not in native spend set"
        }
        ("CORE_VAULT", ThresholdSigErrorDetail::SuiteNotRegistered) => {
            "CORE_VAULT suite not registered"
        }
        ("CORE_VAULT", ThresholdSigErrorDetail::KeyBindingMismatch) => {
            "CORE_VAULT key binding mismatch"
        }
        ("CORE_VAULT", ThresholdSigErrorDetail::SignatureInvalid) => "CORE_VAULT signature invalid",
        ("CORE_VAULT", ThresholdSigErrorDetail::ThresholdNotMet) => "CORE_VAULT threshold not met",
        _ => context,
    }
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn validate_threshold_sig_spend_q(
    keys: &[[u8; 32]],
    threshold: u8,
    ws: &[WitnessItem],
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    block_height: u64,
    context: &'static str,
    cache: &mut SighashV1PrehashCache<'_>,
    sig_queue: Option<&mut SigCheckQueue>,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
) -> Result<(), TxError> {
    let default_rp = DefaultRotationProvider;
    let default_reg = SuiteRegistry::default_registry();
    let rp: &dyn RotationProvider = rotation.unwrap_or(&default_rp);
    let reg = registry.unwrap_or(&default_reg);
    let mut sig_queue = sig_queue;

    if ws.len() != keys.len() {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "witness slot assignment mismatch",
        ));
    }

    let native_spend = rp.native_spend_suites(block_height);
    let mut valid: u8 = 0;
    let queue_mark = sig_queue.as_ref().map(|queue| queue.mark());

    let result = (|| -> Result<(), TxError> {
        for i in 0..keys.len() {
            let w = &ws[i];
            if w.suite_id == SUITE_ID_SENTINEL {
                if !w.pubkey.is_empty() || !w.signature.is_empty() {
                    return Err(TxError::new(
                        ErrorCode::TxErrParse,
                        "SENTINEL witness must be keyless",
                    ));
                }
                continue;
            }

            let params = spend_suite_params(
                w,
                &native_spend,
                reg,
                threshold_sig_error_message(
                    context,
                    ThresholdSigErrorDetail::SuiteNotInNativeSpendSet,
                ),
                threshold_sig_error_message(context, ThresholdSigErrorDetail::SuiteNotRegistered),
            )?;
            validate_witness_item_lengths(w, params)?;

            verify_mldsa_key_and_sig_q(
                w,
                keys[i],
                input_index,
                input_value,
                chain_id,
                cache,
                reg,
                &mut sig_queue,
                TxError::new(
                    ErrorCode::TxErrSigInvalid,
                    threshold_sig_error_message(
                        context,
                        ThresholdSigErrorDetail::KeyBindingMismatch,
                    ),
                ),
                TxError::new(
                    ErrorCode::TxErrSigInvalid,
                    threshold_sig_error_message(context, ThresholdSigErrorDetail::SignatureInvalid),
                ),
            )?;
            valid = valid.saturating_add(1);
        }
        if valid < threshold {
            return Err(TxError::new(
                ErrorCode::TxErrSigInvalid,
                threshold_sig_error_message(context, ThresholdSigErrorDetail::ThresholdNotMet),
            ));
        }
        Ok(())
    })();

    match result {
        Ok(()) => Ok(()),
        Err(err) => {
            if let (Some(mark), Some(queue)) = (queue_mark, sig_queue) {
                queue.rollback_to(mark);
            }
            Err(err)
        }
    }
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn verify_mldsa_key_and_sig_q(
    w: &WitnessItem,
    expected_key_id: [u8; 32],
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    cache: &mut SighashV1PrehashCache<'_>,
    registry: &SuiteRegistry,
    sig_queue: &mut Option<&mut SigCheckQueue>,
    key_binding_error: TxError,
    invalid_sig_error: TxError,
) -> Result<(), TxError> {
    if sha3_256(&w.pubkey) != expected_key_id {
        return Err(key_binding_error);
    }
    let (crypto_sig, sighash_type) = extract_crypto_sig_and_sighash(w)?;
    let digest =
        sighash_v1_digest_with_cache(cache, input_index, input_value, chain_id, sighash_type)?;
    queue_or_verify_signature(
        w.suite_id,
        &w.pubkey,
        crypto_sig,
        digest,
        registry,
        sig_queue,
        invalid_sig_error,
    )
}

#[cfg(test)]
#[path = "spend_verify_tests/mod.rs"]
mod spend_verify_tests;
